core/core_test.go

```diff
@@ -75,6 +82,190 @@ func TestTranscode(t *testing.T) {
+	n.PMSessions[md.ManifestID][sessionIDs[0]] = true
+	n.PMSessions[md.ManifestID][sessionIDs[1]] = true
+	n.pmSessionsMutex.Unlock()
+
+	// Need to use this because the order of the string slice used to call this function
+	// later cannot be guaranteed, since they are coming from a map's keys.
+	argsMatcher := mock.MatchedBy(func(IDs []string) bool {
+		sort.Strings(sessionIDs)
+		sort.Strings(IDs)
+
+		return reflect.DeepEqual(sessionIDs, IDs)
+	})
+
+	recipient.On("RedeemWinningTickets", argsMatcher).Return(nil)
+
+	_, err := n.sendToTranscodeLoop(md, ss)
+	require.Nil(err)
+	waitForTranscoderLoopTimeout(n, md.ManifestID)
+
+	recipient.AssertExpectations(t)
+}
+
+func TestTranscodeLoop_GivenNoPMSession_DoesntTryToRedeem(t *testing.T) {
+	recipient := new(pm.MockRecipient)
+	//Set up the node
+	drivers.NodeStorage = drivers.NewMemoryDriver("")
+	db, _ := common.InitDB("file:TestTranscode?mode=memory&cache=shared")
+	defer db.Close()
+	seth := &eth.StubClient{}
+	tmp, _ := ioutil.TempDir("", "")
+	n, _ := NewLivepeerNode(seth, tmp, db)
+	n.Recipient = recipient
+	defer os.RemoveAll(tmp)
+	ffmpeg.InitFFmpeg()
+	ss := StubSegment()
+	md := &SegTranscodingMetadata{Profiles: videoProfiles}
+	n.Transcoder = NewLocalTranscoder(tmp)
+	transcodeLoopTimeout = 100 * time.Millisecond
+	require := require.New(t)
+	recipient.On("RedeemWinningTickets", mock.Anything).Return(nil)
+
+	_, err := n.sendToTranscodeLoop(md, ss)
+	require.Nil(err)
+	waitForTranscoderLoopTimeout(n, md.ManifestID)
+
+	recipient.AssertNotCalled(t, "RedeemWinningTickets", mock.Anything)
+}
+
+func TestTranscodeLoop_GivenRedeemError_ErrorLogIsWritten(t *testing.T) {
+	recipient := new(pm.MockRecipient)
+	//Set up the node
+	drivers.NodeStorage = drivers.NewMemoryDriver("")
+	db, _ := common.InitDB("file:TestTranscode?mode=memory&cache=shared")
+	defer db.Close()
+	seth := &eth.StubClient{}
+	tmp, _ := ioutil.TempDir("", "")
+	n, _ := NewLivepeerNode(seth, tmp, db)
+	n.Recipient = recipient
+	defer os.RemoveAll(tmp)
+	ffmpeg.InitFFmpeg()
+	ss := StubSegment()
+	md := &SegTranscodingMetadata{Profiles: videoProfiles}
+	n.Transcoder = NewLocalTranscoder(tmp)
+	transcodeLoopTimeout = 100 * time.Millisecond
+	require := require.New(t)
+
+	sessionID := "some session ID"
+	n.pmSessionsMutex.Lock()
+	n.PMSessions[md.ManifestID] = make(map[string]bool)
+	n.PMSessions[md.ManifestID][sessionID] = true
+	n.pmSessionsMutex.Unlock()
+	recipient.On("RedeemWinningTickets", mock.Anything).Return(fmt.Errorf("some error"))
+
+	errorLogsBefore := glog.Stats.Error.Lines()
+	_, err := n.sendToTranscodeLoop(md, ss)
+	require.Nil(err)
+	waitForTranscoderLoopTimeout(n, md.ManifestID)
+	errorLogsAfter := glog.Stats.Error.Lines()
+
+	assert.Equal(t, int64(1), errorLogsAfter-errorLogsBefore)
+}
+
+func waitForTranscoderLoopTimeout(n *LivepeerNode, m ManifestID) {
+	for i := 0; i < 3; i++ {
+		time.Sleep(transcodeLoopTimeout * 2)
+		segChan := getSegChan(n, m)
+		if segChan == nil {
```

Comment:
Should this function *just* wait for the duration of `transcodeLoopTimeout`? At the moment, this function also defines an additional break condition for when the transcode loop is cleaned up (i.e. `segChan == nil`). I feel like this function shouldn't have to know anything about that condition and instead the test should wait for the duration of `transcodeLoopTimeout` and then confirm that the clean up happened. In that case, the rest of this function could be omitted and we could just use `time.Sleep(transcodeLoopTimeout)`

Reply:
I thought so initially, but in practice waiting that long isn't sufficient - the test would pass/fail inconsistently. It even failed for me when I waited `transcodeLoopTimeout * 2`. I prefer this approach to avoid any chance of flaky tests, while paying the cost of some extra coupling between the test and the implementation.

Wdyt?

Reply:
Would it still be flaky if we waited `transcodeLoopTimeout` multiplied by a larger constant like 5?  I do agree that avoiding flaky tests would be preferable

Reply:
Even if you wait 5x the wait time, you're still running the risk of the test failing rarely. I think it's best to have a slight fragile test in terms of its assertion, for the benefit of not making our tests significantly longer than they need to be + we get less flakiness. We just need to be ok with the compromise on the assertion.

I still feel like the current approach is the lesser evil.

Wdyt?

Reply:
Later on we might want to consider stubbing out the context creation itself (rather than overriding the `transcodeLoopTimeout`) in order to gain more control of behavior during testing. We can use a plain cancellable context for testing. Something like:

```golang
// default implementation
var transcodeLoopContext = func() (ctx context.Context, cancel context.CancelFunc) {
  return context.WithTimeout(context.Background(), transcodeLoopTimeout) 
}

// during testing

stubCtx, stubCancel = context.WithCancel(context.Background()) 
stubTranscodeLoopContext = func() {
  return stubCtx, stubCancel
}
transcodeLoopContext = stubTranscodeLoopContext
...
sendToTranscodeLoop(md, ss)
stubCancel()
assert(getSegChan(n, m) == nil)
```

This would give us the opportunity to test other "hard-to-reach" behaviors such as ensuring the `cancel` function is called after each segment.

This mitigates the timing sensitivity from most places in the testing code; about the only place we'd need to be timing-sensitive is when testing the default `transcodeLoopContext` function. This pattern isn't used anywhere else in the codebase yet, but there are a few other places where it'd be useful as well, eg during tests for O discovery.

Reply:
I love the idea! I'll add a note to introduce this refactor, ideally in the time between MVS and SS.

Reply:
@j0sh added an issue: https://github.com/livepeer/go-livepeer/issues/660